smartcontract/service/native/cross_chain/cross_chain_manager/utils.go

```diff
@@ -0,0 +1,147 @@
+}
+
+func checkDoneTx(native *native.NativeService, txHash []byte, chainID uint64) error {
+	contract := utils.CrossChainContractAddress
+	chainIDBytes, err := utils.GetUint64Bytes(chainID)
+	if err != nil {
+		return fmt.Errorf("checkDoneTx, get chainIDBytes error: %v", err)
+	}
+	value, err := native.CacheDB.Get(utils.ConcatKey(contract, []byte(DONE_TX), chainIDBytes, txHash))
+	if err != nil {
+		return fmt.Errorf("checkDoneTx, native.CacheDB.Get error: %v", err)
+	}
+	if value != nil {
+		return fmt.Errorf("checkDoneTx, tx already done")
+	}
+	return nil
+}
+
+func putRequest(native *native.NativeService, txHash []byte, chainID uint64, request []byte) error {
+	contract := utils.CrossChainContractAddress
+	chainIDBytes, err := utils.GetUint64Bytes(chainID)
+	if err != nil {
+		return fmt.Errorf("putRequest, get chainIDBytes error: %v", err)
+	}
+	utils.PutBytes(native, utils.ConcatKey(contract, []byte(REQUEST), chainIDBytes, txHash), request)
+	return nil
+}
+
+func MakeFromOntProof(native *native.NativeService, params *CreateCrossChainTxParam) error {
+	//record cross chain tx
+	txHash := native.Tx.Hash()
+	merkleValue := &ccom.MakeTxParam{
+		TxHash:              txHash.ToArray(),
+		FromContractAddress: native.ContextRef.CallingContext().ContractAddress[:],
+		ToChainID:           params.ToChainID,
+		ToContractAddress:   params.ToContractAddress,
+		Method:              params.Method,
+		Args:                params.Args,
+	}
+	sink := common.NewZeroCopySink(nil)
+	merkleValue.Serialization(sink)
+	err := putRequest(native, merkleValue.TxHash, params.ToChainID, sink.Bytes())
+	if err != nil {
+		return fmt.Errorf("MakeFromOntProof, putRequest error:%s", err)
+	}
+	native.PushCrossState(sink.Bytes())
+	chainIDBytes, err := utils.GetUint64Bytes(params.ToChainID)
+	if err != nil {
+		return fmt.Errorf("MakeFromOntProof, get chainIDBytes error: %v", err)
+	}
+	key := hex.EncodeToString(utils.ConcatBytes([]byte(REQUEST), chainIDBytes, merkleValue.TxHash))
+	args := hex.EncodeToString(params.Args)
+	notifyMakeFromOntProof(native, hex.EncodeToString(merkleValue.TxHash), params.ToChainID, key,
+		hex.EncodeToString(merkleValue.FromContractAddress), args)
+	return nil
+}
+
+func VerifyToOntTx(native *native.NativeService, proof []byte, fromChainid uint64, header *ccom.Header) (*ccom.ToMerkleValue, error) {
+	v, err := merkle.MerkleProve(proof, header.CrossStatesRoot)
+	if err != nil {
+		return nil, fmt.Errorf("VerifyToOntTx, merkle.MerkleProve verify merkle proof error: %v", err)
+	}
+
+	s := common.NewZeroCopySource(v)
+	merkleValue := new(ccom.ToMerkleValue)
+	if err := merkleValue.Deserialization(s); err != nil {
+		return nil, fmt.Errorf("VerifyToOntTx, deserialize merkleValue error:%s", err)
+	}
+
+	//record done cross chain tx
+	err = checkDoneTx(native, merkleValue.TxHash, fromChainid)
+	if err != nil {
+		return nil, fmt.Errorf("VerifyToOntTx, checkDoneTx error:%s", err)
+	}
+	err = putDoneTx(native, merkleValue.TxHash, fromChainid)
```

Comment:
so each tx can create no more than one cross chain transaction? but I do not find any limit check when calling `CreateCrossChainTx`

Reply:
fixed